_posts/2022/2022-01-17-enqueue-jobs-quickly-with-sidekiq-bulk.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "Enqueue Jobs Quickly with Sidekiq’s Bulk Features"
+description: "From thousands of network calls, to one."
+layout: article
+category: ruby
+image:
+  base: '2022/enqueue-jobs-quickly-with-sidekiq-bulk'
+  alt: "Martial Arts on the beach"
+  credit: "Thao Le Hoang"
+  source: "https://unsplash.com/photos/Xl-ilWBKJNk"
+
+---
+
+We often have to enqueue lots of the same jobs at the same time. Perhaps a data migration for all of one kind of object? Or a data processing step that will run faster if it's parallelised?
+
+Our focus is often on the potentially large queues or the elapsed time it takes to churn through all the work. There is another task that we have to consider. Enqueing all that work in the first place.
+
+
+## Instead of…
+
+…looping through lots of Active Record objects:
+
+```ruby
+Customer.all.each do |customer|
+  DeliverEmailJob.perform_async(customer)
+end
+```
+
+...or better, if you have more than a few hundred jobs to enqueue:
+
+```ruby
+Customer.find_in_batches do |customer|
+  DeliverEmailJob.perform_async(customer)
+end
+```
+
+
+## Use…
+
+…Sidekiq directly:
+
+```ruby
+ids = Customer.all.pluck(:id)
+array_of_args = ids.map { |x| [x] }
+DeliverEmailJob.perform_bulk(array_of_args)
+```
+
+Or, if using a Sidekiq version before 6.3.0:
+
+```ruby
+ids = Customer.all.pluck(:id)
+array_of_args = ids.map { |x| [x] }
+Sidekiq::Client.push_bulk('class' => DeliverEmailJob, 'args' => array_of_args)
+```
+
+
+## Why?
+
+Primarily this minimizes round trips to Redis. Instead of making an individual call for each enqueing action across the network, you make _one_. This is a big deal even when using code as well optimised for speed as Redis.
+
+This can also have the benefit (on top of passing scaler `id`s to jobs) of minimising the amount of memory used. By `pluck`ing `id`s rather than looping over arrays of Active Record models you are simply dealing with less complex objects which take up less memory.
```

Comment:
```suggestion
This can also have the benefit of minimising the amount of memory used. By `pluck`ing `id`s, or using the `ids` method as shown, rather than looping over arrays of Active Record models you are simply dealing with less complex objects which take up less memory.
```